### ISSUE TITLE
Backward_Relevance_cpu implementation on new layers

### ISSUE THREAD

**Issue by oscmansan** [sebastian-lapuschkin/lrp_toolbox#10]
Hi,

I want to try the toolbox with a ResNet-50 architecture. The problem is that it uses some layers that were not present in the Caffe version of the toolbox. In particular, the _BatchNorm_, _Bias_ and _Scale_ layers are missing. I've already copied them from the newest Caffe version, but I'm missing the implementation of `Backward_Relevance_cpu` in these layers. Do you have any plans on implementing them? If not, any pointers on how to implement it?

**Comment by sebastian-lapuschkin** [comment]
Hi Oscar, 

there is, unfortunately very little activity in the Caffe front, since the main contributors are quite busy otherwise. My collaborator [Alexander Binder](https://istd.sutd.edu.sg/people/faculty/alexander-binder) should have experimental code for ResNets and BatchNorm layers on his workstation somewhere.  Maybe contacting him will help you out.

General pointers for implementing the layers on your own:
BatchNorm, Bias and Scale all realize affine transformations.  The learned bias/mean/shift parameters adding/subtracting to/from the hidden representations of your inputs should be treated as regular neurons, e.g. they will affect the output neuron activations of the current layer and thus should absorb some quantity of relevance. 

(I will use latex-syntax for math from here on).
For the BatchNorm layer, let's assume you populate your variance $\sigma_B^2$ and mean $\mu_B$ parameters based on your training population (so you can run single-sample batches). 

Use [this](https://kratzert.github.io/2016/02/12/understanding-the-gradient-flow-through-the-batch-normalization-layer.html):
![image](https://user-images.githubusercontent.com/17984624/36423147-0f53e9ce-163f-11e8-9643-ea7d3d1369e4.png)

The normalizing and scale-and-shift steps can be structured as a sequence of four shift and normalization steps, only executing one step at a time. Look at the syntax in [this](http://openaccess.thecvf.com/content_ICCV_2017_workshops/papers/w23/Lapuschkin_Understanding_and_Comparing_ICCV_2017_paper.pdf), Formula (2).
Here, $z_{ij}$ is the forward contribution from input neuron $i$ to output neuron $j$ and $z_j$ are the pooled forward contributions at output $j$ (i.e, $z_j$ is the output activation).

For the scaling steps within the BatchNorm layer, you have
$z_i = cx_i$. Note that this is a component-wise operation, therefore $z_{ij} = 0$ for all $i \neq j$.
This is why I write $z_i$ only. Also, as a result of the component-wise application of the scaling:
$z_j = z_i$ and in terms of Formulas (2) and (3) in above linked paper you have $R_i = R_j$ if i equals j.
Long story short: do nothing for the scaling step.

For the (mean) shift steps, treat the shift parameter as a bias neuron $b$. Note that the operation is again component-wise, therefore only one index $i$ again, with $i=j$:
$z_j = z_i + z_{b_i}$ or just $z_j = x_i + b_i$. Consecutively, for attributing relevance to the inputs:
$R_{x_i} = \frac{x_i}{x_i + b_i}R_j$. Note that the shifting bias would absorb some of the relevance, since it affects the forward pass as a constant source of 'activation energy': $R_{b_i} = \frac{b_i}{x_i + b_i}R_j$. It is up to you, whether you want to model the relevance attribution to the bias explicitly or not. $R_{x_i}$ should suffice for most cases.

I hope that helps :)

**Comment by oscmansan** [comment]
Hi Sebastian,

Thank you very much for taking the time to write such a complete response :). I've already contacted your colleague asking him for the experimental code, but I'm going to try to implement it myself since after your explanation it actually seems a feasible task.

I think I understood your implementation pointers for the _BatchNorm_ and _Scale_ layers. However, I've noticed that the ResNet architecture also uses the _EltwiseLayer_ with a SUM operation, which is present in the Caffe version of the toolbox but it doesn't have the `Backward_Relevance_cpu` implementation either. I've come up with a way of attributing relevance to the inputs depending on the relevance of the outputs (also in latex notation): let $i_{1}$ and $i_{2}$ be two neurons belonging to two inputs of the layer and let $j$ be a neuron of the output such that $y_{j} = x_{i_{1}} + x_{i_{2}}$, then $R_{i_{1}} = \frac{x_{i_{1}}}{x_{i_{1}} + x_{i_{2}}}R_{j}$ and $R_{i_{2}} = \frac{x_{i_{2}}}{x_{i_{1}} + x_{i_{2}}}R_{j}$. Do you think this would be a correct implementation?

**Comment by sebastian-lapuschkin** [comment]
Hi Oscar, 

looks good.
Have a look at the numerics in this layer, though, especially with the inputs of both summands canceling each other out. Consider adding a small epsilon to the denominator, or use the alpha-beta rule you can find in our papers.
